cmd/runner/main.go

```diff
@@ -53,16 +70,294 @@ func run(args []string) error {
+	if azClientSecret := os.Getenv("AZURE_CLIENT_SECRET"); azClientSecret != "" {
+		if creds, err = azidentity.NewEnvironmentCredential(nil); err != nil {
+			return fmt.Errorf("failed to load Azure credentials from environment: %w", err)
+		}
+	} else {
+		// We use "ManagedIdentity" instead of just "Default" because the default
+		// timeout is too low in azidentity.NewDefaultAzureCredentials, so it times out
+		// and fails to run.
+		azClientID := os.Getenv("AZURE_CLIENT_ID")
+		logger.Info("Loading user managed credentials", zap.String("client_id", azClientID))
+		if creds, err = azidentity.NewManagedIdentityCredential(&azidentity.ManagedIdentityCredentialOptions{
+			ID: azidentity.ClientID(azClientID),
+		}); err != nil {
+			return fmt.Errorf("failed to load Azure credentials: %w", err)
+		}
+	}
+
+	blobClient, err := azblob.NewClient(creds, *azStorageAccount)
+	if err != nil {
+		return fmt.Errorf("failed to init blob client: %w", err)
+	}
+
+	h := handler{
+		blob:                     blobClient,
+		sourcePortfolioContainer: *azSourcePortfolioContainer,
+		destPortfolioContainer:   *azDestPortfolioContainer,
+		reportContainer:          *azReportContainer,
+	}
+
+	validTasks := map[task.Type]func(context.Context) error{
+		task.ProcessPortfolio: toRunFn(processPortfolioReq, h.processPortfolio),
+		task.CreateReport:     toRunFn(createReportReq, h.createReport),
+	}
+
+	taskType := task.Type(os.Getenv("TASK_TYPE"))
+	if taskType == "" {
+		return errors.New("no TASK_TYPE given")
+	}
+
+	taskFn, ok := validTasks[taskType]
+	if !ok {
+		return fmt.Errorf("unknown task type %q", taskType)
+	}
+
+	logger.Info("running PACTA task", zap.String("task_type", string(taskType)))
+
+	if err := taskFn(ctx); err != nil {
+		return fmt.Errorf("error running task: %w", err)
+	}
+
+	logger.Info("ran PACTA task successfully", zap.String("task_type", string(taskType)))
+
+	return nil
+}
+
+type Blob interface {
+	ReadBlob(ctx context.Context, uri string) (io.ReadCloser, error)
+	WriteBlob(ctx context.Context, uri string, r io.Reader) error
+	Scheme() blob.Scheme
+}
+
+type handler struct {
+	blob                     Blob
+	sourcePortfolioContainer string
+	destPortfolioContainer   string
+	reportContainer          string
+}
+
+func processPortfolioReq() (*task.ProcessPortfolioRequest, error) {
+	rawAssetIDs := os.Getenv("ASSET_IDS")
+	if rawAssetIDs == "" {
+		return nil, errors.New("no ASSET_IDS given")
+	}
+
+	var assetIDs []string
+	if err := json.NewDecoder(strings.NewReader(rawAssetIDs)).Decode(&assetIDs); err != nil {
+		return nil, fmt.Errorf("failed to load asset IDs: %w", err)
+	}
+
+	return &task.ProcessPortfolioRequest{
+		AssetIDs: assetIDs,
+	}, nil
+}
+
+func (h *handler) uploadDirectory(ctx context.Context, dirPath, container string) error {
+	base := filepath.Base(dirPath)
+
+	return filepath.WalkDir(dirPath, func(path string, info fs.DirEntry, err error) error {
```

Comment:
I'm all for not prematurely optimizing, but IIUC, we know that a given run is likely to operate over a large number of assets (particularly, portfolio group reports). I think we will want to parallelize this - using walk for enumeration of tasks, and then a helper to do `UploadAllBlobs`

Reply:
Good to know, I'm still going to let it ride and see what actual performance looks like. For example, if processing latency is 10x upload latency, this will be a bit of work and complexity for no additional gain.

An `UploadAllBlobs` function needs to make some decisions about how many goroutines to run in parallel for the upload, which requires some experimenting, so I'll save all that until we know we need it.